ml-algorithms/src/main/java/org/opensearch/ml/engine/memory/ConversationIndexMemory.java

```diff
@@ -0,0 +1,215 @@
+        String memoryMessageIndexName,
+        String conversationId,
+        MLMemoryManager memoryManager
+    ) {
+        this.client = client;
+        this.mlIndicesHandler = mlIndicesHandler;
+        this.memoryMetaIndexName = memoryMetaIndexName;
+        this.memoryMessageIndexName = memoryMessageIndexName;
+        this.conversationId = conversationId;
+        this.memoryManager = memoryManager;
+    }
+
+    @Override
+    public String getType() {
+        return TYPE;
+    }
+
+    @Override
+    public void save(String id, Message message) {
+        this.save(id, message, ActionListener.wrap(r -> { log.info("saved message into {} memory, session id: {}", TYPE, id); }, e -> {
+            log.error("Failed to save message to memory", e);
+        }));
+    }
+
+    @Override
+    public void save(String id, Message message, ActionListener listener) {
+        mlIndicesHandler.initMemoryMessageIndex(ActionListener.wrap(created -> {
+            if (created) {
+                IndexRequest indexRequest = new IndexRequest(memoryMessageIndexName);
+                ConversationIndexMessage conversationIndexMessage = (ConversationIndexMessage) message;
+                XContentBuilder builder = XContentBuilder.builder(XContentType.JSON.xContent());
+                conversationIndexMessage.toXContent(builder, ToXContent.EMPTY_PARAMS);
+                indexRequest.source(builder);
+                client.index(indexRequest, listener);
+            } else {
+                listener.onFailure(new RuntimeException("Failed to create memory message index"));
+            }
+        }, e -> { listener.onFailure(new RuntimeException("Failed to create memory message index")); }));
+    }
+
+    public void save(Message message, String parentId, Integer traceNum, String action) {
+        this.save(message, parentId, traceNum, action, ActionListener.<CreateInteractionResponse>wrap(r -> {
+            log
+                .info(
+                    "saved message into memory {}, parent id: {}, trace number: {}, interaction id: {}",
+                    conversationId,
+                    parentId,
+                    traceNum,
+                    r.getId()
+                );
+        }, e -> { log.error("Failed to save interaction", e); }));
+    }
+
+    public void save(Message message, String parentId, Integer traceNum, String action, ActionListener listener) {
+        ConversationIndexMessage msg = (ConversationIndexMessage) message;
+        memoryManager
+            .createInteraction(conversationId, msg.getQuestion(), null, msg.getResponse(), action, null, parentId, traceNum, listener);
+    }
+
+    @Override
+    public void getMessages(String id, ActionListener listener) {
+        SearchRequest searchRequest = new SearchRequest();
+        searchRequest.indices(memoryMessageIndexName);
+        SearchSourceBuilder sourceBuilder = new SearchSourceBuilder();
+        sourceBuilder.size(10000);
```

Comment:
Should we keep this as a default value?

Reply:
At this moment, we have it as a default value.

Reply:
[nit] Can we set this value as a static variable in that case?